lib/logstash/codecs/fluent.rb

```diff
@@ -145,4 +148,14 @@ def generate_event(map, fluent_time, tag)
     event
   end
 
+  def map_timestamp(event_hash)
```

Comment:
There is still an issue with complex events that may have multiple hierarchies, and the arrays will also need to be recursed in the same way the hashes are. 

Given [this comment](https://github.com/logstash-plugins/logstash-codec-fluent/pull/30/files#diff-46e627c46cbd94c095fec7ceca988ead0ed7d15c52c170134906f862227d7409R80) about not mutating the original data, maybe we could implement `map_timestamp` like this?

```suggestion
  def normalize_timestamps(o)
    case o
    when Hash
      o.inject({}){|c, (k, v)| c[k] = map_timestamp(v); c}
    when Array
      o.map{|i| map_timestamp(i)}
    when LogStash::Timestamp
      o.to_iso8601
    else
      o
    end
  end
```
  In which case, our `@on_event.call` method would look like:

```
@on_event.call(event, @packer.pack([tag, epochtime, normalize_timestamps(data)]))
```

Reply:
I believe with your suggestion we need to opt out the `map_timestamp` method and pushed with [this commit](https://github.com/logstash-plugins/logstash-codec-fluent/pull/30/commits/e0d7a0e5948d94fb5d5c76f7d8af6de73ac66621).